mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -57,9 +57,53 @@ struct ConcatOptimization : public OpRewritePattern<tosa::ConcatOp> {
   }
 };
 
+struct ConcatFolding : public OpRewritePattern<tosa::ConcatOp> {
+  using OpRewritePattern<tosa::ConcatOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ConcatOp op,
+                                PatternRewriter &rewriter) const override {
+    // Fold consecutive concats on the same axis into a single op.
+    uint64_t axis = op.getAxis();
+
+    // Keep track of the operands so we are able to construct a new concat
+    // later. Conservatively assume that we double the number of operands when
+    // folding
+    SmallVector<Value, 8> concatOperands;
+    concatOperands.reserve(2 * op->getNumOperands());
+
+    // Find all operands that are foldable concats
+    bool canFold = false;
+    for (Value operand : op->getOperands()) {
+      concatOperands.emplace_back(operand);
+
+      auto producer = dyn_cast_or_null<ConcatOp>(operand.getDefiningOp());
+      if (!producer)
+        continue;
+
+      // Foldable if axis is the same
+      if (axis != producer.getAxis())
+        continue;
+
+      // Replace the original operand with all incoming operands
+      canFold = true;
```

Comment:
Should we also check that the producer has no other uses?

Reply:
There's no need to do that. If there are other users, it will not get DCE'd but at least we optimized one of them and enable more kernel matching & chaining later